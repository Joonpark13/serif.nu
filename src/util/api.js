/* eslint-disable import/prefer-default-export */
export function searchURL(query) {
  return `${process.env.API_URL}/search?term_id=4720&search_query=${query}`;
}

export function sectionsURL(termId, schoolAbbv, subjectAbbv, courseAbbv) {
  return `${process.env.API_URL}/sections?term_id=${termId}&school_abbv=${schoolAbbv}&subject_abbv=${subjectAbbv}&course_abbv=${courseAbbv}`;
}
