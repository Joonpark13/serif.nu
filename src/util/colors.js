export const northwesternPurple = '#4E2A84';
export const northwesternBrightOrange = '#FFC520';
export const northwesternRichBlack10 = '#D8D6D6';
