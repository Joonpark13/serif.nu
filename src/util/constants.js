/* eslint-disable import/prefer-default-export */
export const CURRENT_TERM = '4730';
