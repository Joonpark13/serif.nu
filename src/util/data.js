import termsData from 'data/terms.json';

/* eslint-disable import/prefer-default-export */
/* istanbul ignore next */
export const currentTerm = termsData.reduce((highestTerm, term) => {
  if (term.id > highestTerm.id) {
    return term;
  }
  return highestTerm;
});
/* eslint-enable import/prefer-default-export */
