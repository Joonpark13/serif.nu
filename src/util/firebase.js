import firebase from 'firebase/app';
import 'firebase/firestore';
import 'firebase/auth';

firebase.initializeApp({
  apiKey: 'AIzaSyDluAG6bnqleqJ3yjPvGdlb3RyZP3U5JIQ',
  authDomain: 'sans-serif-northwestern.firebaseapp.com',
  databaseURL: 'https://sans-serif-northwestern.firebaseio.com',
  projectId: 'sans-serif-northwestern',
  storageBucket: 'sans-serif-northwestern.appspot.com',
  messagingSenderId: '1078100635889',
});

export const db = firebase.firestore();
export const auth = firebase.auth();
