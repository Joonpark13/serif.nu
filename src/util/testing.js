/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import * as reactRedux from 'react-redux';
import { shallow } from 'enzyme';
import * as useSelector from './use-selector';

jest.mock('./use-selector');

// Helper function that returns a function which creates enzyme wrappers with default props
export function wrapperCreator(Component, defaultProps = {}) {
  let classes;
  return props => shallow(<Component classes={classes} {...defaultProps} {...props} />);
}

/* istanbul ignore next */
export function mockUseSelector(...args) {
  useSelector.default.mockReset();
  args.forEach(arg => useSelector.default.mockReturnValueOnce(arg));

  return useSelector.default;
}

/* istanbul ignore next */
export function mockUseDispatch() {
  const dispatchSpy = jest.fn();
  reactRedux.useDispatch.mockReturnValue(dispatchSpy);
  return dispatchSpy;
}
