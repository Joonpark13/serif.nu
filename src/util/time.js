export function formatMinute(minute) {
  if (minute < 10) return `0${minute}`;
  return `${minute}`;
}

export function formatTime(time, includeAmPm = true) {
  if (time.hour === 0) return `12:${formatMinute(time.minute)}${includeAmPm ? ' AM' : ''}`;
  if (time.hour < 12) return `${time.hour}:${formatMinute(time.minute)}${includeAmPm ? ' AM' : ''}`;
  if (time.hour === 12) return `12:${formatMinute(time.minute)}${includeAmPm ? ' PM' : ''}`;
  return `${time.hour - 12}:${formatMinute(time.minute)}${includeAmPm ? ' PM' : ''}`;
}

export function getFormattedClassEvent(event) {
  if (event.dow === 'TBA' || event.start === 'TBA' || event.end === 'TBA') {
    return 'TBA';
  }
  return `${formatTime(event.start, false)} - ${formatTime(event.end, false)}`;
}

export function getFormattedClassSchedule(event) {
  if (event.dow === 'TBA' || event.start === 'TBA' || event.end === 'TBA') {
    return 'TBA';
  }
  const dowStr = event.dow.join('');
  return `${dowStr} ${formatTime(event.start, true)} - ${formatTime(event.end, true)}`;
}

export function getDurationInHours(event) {
  const hoursDiff = event.end.hour - event.start.hour;
  const minutesDiff = event.end.minute - event.start.minute;
  return hoursDiff + minutesDiff / 60;
}

export function isBefore(time1, time2, orEqual = false) {
  if (time1.hour < time2.hour) {
    return true;
  }
  if (time1.hour === time2.hour) {
    if (orEqual) return time1.minute <= time2.minute;
    return time1.minute < time2.minute;
  }
  return false;
}

export function overlaps(event1, event2) {
  if (event1.dow !== event2.dow) {
    return false;
  }
  return !(isBefore(event1.end, event2.start) || isBefore(event2.end, event1.start));
}
