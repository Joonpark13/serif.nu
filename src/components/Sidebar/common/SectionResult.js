import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Typography from '@material-ui/core/Typography';
import { getFormattedClassSchedule } from 'util/time';

export const styles = {
  sectionTitle: {
    fontWeight: 'bold',
  },
};

function SectionResult({ addSection, sectionHover, sectionHoverOff, section, disabled, classes }) {
  const isUnscheduled = section.schedule.some(
    schedule => schedule.dow === 'TBA' || schedule.start === 'TBA' || schedule.end === 'TBA',
  );
  return (
    <ListItem
      key={section.id}
      button
      onClick={() => addSection(section)}
      onMouseEnter={() => sectionHover(section)}
      onMouseLeave={sectionHoverOff}
      disabled={disabled || isUnscheduled}
    >
      <ListItemText>
        <Typography variant="h6" className={classes.sectionTitle}>
          {`Section ${section.sectionNumber}`}
        </Typography>

        {section.topic && <Typography>{section.topic}</Typography>}

        {section.schedule.map((scheduleObj, index) => (
          /* eslint-disable react/no-array-index-key */
          <Typography color={isUnscheduled ? 'error' : undefined} key={index}>
            {getFormattedClassSchedule(scheduleObj)}
          </Typography>
          /* eslint-enable react/no-array-index-key */
        ))}

        <Typography>{section.schedule[0].location}</Typography>

        <Typography>
          {section.instructors.map((teacher, idx) => (
            `${teacher}${idx !== section.instructors.length - 1 ? ', ' : ''}`))
          }
        </Typography>
      </ListItemText>
    </ListItem>
  );
}

SectionResult.propTypes = {
  addSection: PropTypes.func.isRequired,
  sectionHover: PropTypes.func.isRequired,
  sectionHoverOff: PropTypes.func.isRequired,
  section: PropTypes.objectOf(PropTypes.any).isRequired, // TODO
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  disabled: PropTypes.bool,
};

SectionResult.defaultProps = {
  disabled: false,
};

export { SectionResult as UnstyledSectionResult };
export default withStyles(styles)(SectionResult);
