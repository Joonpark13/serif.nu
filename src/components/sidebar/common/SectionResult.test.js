import ListItem from '@material-ui/core/ListItem';
import Typography from '@material-ui/core/Typography';
import { wrapperCreator } from 'util/testing';
import * as timeUtils from 'util/time';
import { UnstyledSectionResult, styles } from './SectionResult';

jest.mock('util/time');

describe('SectionResult', () => {
  const formattedSchedule = 'MWF 10 - 12ish';
  timeUtils.getFormattedClassSchedule.mockReturnValue(formattedSchedule);

  const section = {
    id: '198732',
    sectionNumber: 20,
    topic: 'Section topic...',
    schedules: [{ location: 'somewhere' }],
    instructors: ['A prof'],
  };
  const defaultProps = {
    addSection: () => {},
    section,
    sectionHover: () => {},
    sectionHoverOff: () => {},
  };
  const getWrapper = wrapperCreator(UnstyledSectionResult, defaultProps, styles);

  it('renders correctly', () => {
    const wrapper = getWrapper();

    expect(wrapper.get(0)).toMatchSnapshot();
  });

  it('renders instructors correctly for multiple instructors', () => {
    const wrapper = getWrapper({ section: { ...section, instructors: ['Prof 1', 'Prof 2'] } });

    expect(wrapper.find(Typography).at(4).prop('children')).toEqual(['Prof 1, ', 'Prof 2']);
  });

  it('adds section when clicked', () => {
    const addSectionMock = jest.fn();
    const wrapper = getWrapper({ addSection: addSectionMock });
    wrapper.find(ListItem).simulate('click');

    expect(addSectionMock).toHaveBeenCalledWith(section);
  });

  it('turns scheduled text to red if section is unscheduled', () => {
    timeUtils.getFormattedClassSchedule.mockReturnValue('TBA');
    timeUtils.isUnscheduled.mockReturnValue(true);
    const unscheduledSection = {
      id: '3',
      sectionNumber: 21,
      schedules: [{
        location: 'Some other building',
        dow: 'TBA',
        start: 'TBA',
        end: 'TBA',
      }],
      instructors: ['Ian Horswill', 'Vincent St-Amour'],
    };
    const wrapper = getWrapper({ section: unscheduledSection });
    const colorProp = wrapper
      .findWhere(
        node => node.is(Typography) && node.prop('children') === 'TBA',
      )
      .prop('color');
    expect(colorProp).toBe('error');
  });

  it('calls sectionHover when clicked', () => {
    const sectionHoverMock = jest.fn();
    const wrapper = getWrapper({ sectionHover: sectionHoverMock });
    wrapper.find(ListItem).simulate('mouseEnter');

    expect(sectionHoverMock).toHaveBeenCalledWith(section);
  });
});
