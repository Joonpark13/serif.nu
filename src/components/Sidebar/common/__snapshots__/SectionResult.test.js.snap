// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionResult renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SectionResult
    addSection={[Function]}
    classes={
      Object {
        "sectionTitle": "",
      }
    }
    disabled={false}
    section={
      Object {
        "id": "198732",
        "instructors": Array [
          "A prof",
        ],
        "schedule": Array [
          Object {
            "location": "somewhere",
          },
        ],
        "sectionNumber": 20,
        "topic": "Section topic...",
      }
    }
    sectionHover={[Function]}
    sectionHoverOff={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "198732",
    "nodeType": "class",
    "props": Object {
      "button": true,
      "children": <WithStyles(ListItemText)>
        <WithStyles(Typography)
          className=""
          variant="h6"
        >
          Section 20
        </WithStyles(Typography)>
        <WithStyles(Typography)>
          Section topic...
        </WithStyles(Typography)>
        <WithStyles(Typography)>
          MWF 10 - 12ish
        </WithStyles(Typography)>
        <WithStyles(Typography)>
          somewhere
        </WithStyles(Typography)>
        <WithStyles(Typography)>
          A prof
        </WithStyles(Typography)>
      </WithStyles(ListItemText)>,
      "disabled": false,
      "onClick": [Function],
      "onMouseEnter": [Function],
      "onMouseLeave": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <WithStyles(Typography)
            className=""
            variant="h6"
          >
            Section 20
          </WithStyles(Typography)>,
          <WithStyles(Typography)>
            Section topic...
          </WithStyles(Typography)>,
          Array [
            <WithStyles(Typography)>
              MWF 10 - 12ish
            </WithStyles(Typography)>,
          ],
          <WithStyles(Typography)>
            somewhere
          </WithStyles(Typography)>,
          <WithStyles(Typography)>
            A prof
          </WithStyles(Typography)>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Section 20",
            "className": "",
            "variant": "h6",
          },
          "ref": null,
          "rendered": "Section 20",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Section topic...",
          },
          "ref": null,
          "rendered": "Section topic...",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "class",
          "props": Object {
            "children": "MWF 10 - 12ish",
            "color": undefined,
          },
          "ref": null,
          "rendered": "MWF 10 - 12ish",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "somewhere",
          },
          "ref": null,
          "rendered": "somewhere",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "A prof",
            ],
          },
          "ref": null,
          "rendered": Array [
            "A prof",
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "198732",
      "nodeType": "class",
      "props": Object {
        "button": true,
        "children": <WithStyles(ListItemText)>
          <WithStyles(Typography)
            className=""
            variant="h6"
          >
            Section 20
          </WithStyles(Typography)>
          <WithStyles(Typography)>
            Section topic...
          </WithStyles(Typography)>
          <WithStyles(Typography)>
            MWF 10 - 12ish
          </WithStyles(Typography)>
          <WithStyles(Typography)>
            somewhere
          </WithStyles(Typography)>
          <WithStyles(Typography)>
            A prof
          </WithStyles(Typography)>
        </WithStyles(ListItemText)>,
        "disabled": false,
        "onClick": [Function],
        "onMouseEnter": [Function],
        "onMouseLeave": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <WithStyles(Typography)
              className=""
              variant="h6"
            >
              Section 20
            </WithStyles(Typography)>,
            <WithStyles(Typography)>
              Section topic...
            </WithStyles(Typography)>,
            Array [
              <WithStyles(Typography)>
                MWF 10 - 12ish
              </WithStyles(Typography)>,
            ],
            <WithStyles(Typography)>
              somewhere
            </WithStyles(Typography)>,
            <WithStyles(Typography)>
              A prof
            </WithStyles(Typography)>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Section 20",
              "className": "",
              "variant": "h6",
            },
            "ref": null,
            "rendered": "Section 20",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Section topic...",
            },
            "ref": null,
            "rendered": "Section topic...",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "class",
            "props": Object {
              "children": "MWF 10 - 12ish",
              "color": undefined,
            },
            "ref": null,
            "rendered": "MWF 10 - 12ish",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "somewhere",
            },
            "ref": null,
            "rendered": "somewhere",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                "A prof",
              ],
            },
            "ref": null,
            "rendered": Array [
              "A prof",
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
